zokrates_core/src/flatten/mod.rs

```diff
@@ -724,6 +724,33 @@ impl<'ast, T: Field> Flattener<'ast, T> {
 
                 FlatExpression::Identifier(sub_bits_be[bitwidth - 1])
             }
+            BooleanExpression::BoolEq(box lhs, box rhs) => {
+                // Wanted: Not((x-y)^2)
+                let x = self.flatten_boolean_expression(symbols, statements_flattened, lhs);
+                let y = self.flatten_boolean_expression(symbols, statements_flattened, rhs);
+
+                let name_x_sub_y = self.use_sym();
```

Comment:
`x - y` is linear so no need to introduce a new variable: `let x_sub_y = FlatExpression::Sub(box x, box y)`